src/daft-logical-plan/src/optimization/rules/extract_window_function.rs

```diff
@@ -0,0 +1,488 @@
+        match expr.as_ref() {
+            Expr::Window(_inner_expr, window_spec) => {
+                result.push((expr.clone(), window_spec.clone()));
+            }
+            _ => {
+                for child in expr.children() {
+                    Self::collect_window_functions(&child, result);
+                }
+            }
+        }
+    }
+
+    fn replace_window_functions(
+        expr: &ExprRef,
+        window_col_mappings: &[(ExprRef, String)],
+    ) -> DaftResult<ExprRef> {
+        let transformed = expr.clone().transform(|e| {
+            for (window_expr, col_name) in window_col_mappings {
+                if Arc::ptr_eq(&e, window_expr) {
+                    return Ok(Transformed::yes(resolved_col(col_name.clone())));
+                }
+            }
+            Ok(Transformed::no(e))
+        })?;
+
+        Ok(transformed.data)
+    }
+}
+
+impl OptimizerRule for ExtractWindowFunction {
+    /// Optimizes a logical plan by extracting window function expressions from projections
+    /// and converting them into explicit Window operations.
+    ///
+    /// Algorithm:
+    /// 1. Find all window functions in a projection
+    /// 2. Group them by their window specifications
+    /// 3. Create a separate Window operation for each group
+    /// 4. Connect these Window operations in a chain
+    /// 5. Replace the original window expressions with column references
+    /// 6. Create a final Project operation with the updated expressions
+    fn try_optimize(&self, plan: Arc<LogicalPlan>) -> DaftResult<Transformed<Arc<LogicalPlan>>> {
+        plan.transform_down(|node| match node.as_ref() {
+            LogicalPlan::Project(project) => {
+                if Self::contains_window_function(project) {
+                    let window_funcs = Self::extract_window_functions(&project.projection);
+
+                    debug_assert!(
+                        !window_funcs.is_empty(),
+                        "Window functions detected but none extracted"
+                    );
+
+                    let window_funcs_grouped_by_spec = window_funcs
+                        .into_iter()
+                        .into_group_map_by(|(_, spec)| spec.clone());
+
+                    let mut window_col_mappings = Vec::new();
+
+                    let mut current_plan = project.input.clone();
+
+                    for (window_spec, window_exprs_for_spec) in window_funcs_grouped_by_spec {
+                        let window_function_exprs = window_exprs_for_spec
+                            .iter()
+                            .map(|(expr, _)| expr.clone())
+                            .collect::<Vec<ExprRef>>();
+
+                        let window_function_exprs_aliased = window_function_exprs
+                            .iter()
+                            .map(|e| {
+                                let semantic_id = e.semantic_id(&current_plan.schema());
```

Comment:
add a quick comment here explaining what semantic ID is used for and that you also do the same thing in spec_mappings so that the expressions in the final project point to the right columns